### PR TITLE
Evaluate migration of `SmallVec` to `min_const_generics`

### LINKED ISSUE #588: Evaluate usage of `min_const_generics` for SmallVec
`min_const_generics` is about to be stabilized. It allows us to have `const N: usize` as a generic parameter in a type.

We should evaluate if we can get rid of our `generic-array` and `typenum` dependencies for `ink_storage::SmallVec`.

### PR DESCRIPTION
**Pull request by cmichi**
Closes #588.

### CODE REVIEW

File: crates/storage/src/lazy/lazy_array.rs
Comment (cmichi) [inline crates/storage/src/lazy/lazy_array.rs:223]: `#![feature(array_map)]` is required to make this work. The issue is that we need to fill the array initially and the array value `CacheCell` does not support `Copy`. So `[Default::default(); N]` does not work.

I tried using our `array-init` dependency, but from my understanding it doesn't support `min_const_generics` yet.

File: crates/storage/src/lazy/lazy_array.rs
Comment (cmichi) [inline crates/storage/src/lazy/lazy_array.rs:248]: Not sure about if I can make this change. Hmm, at least everything builds and tests run.

If I include the `.as_slice()` I need `#![feature(array_methods)]`, otherwise I get build errors.

File: crates/storage/src/lazy/cache_cell.rs
Comment (Robbepop) [inline crates/storage/src/lazy/cache_cell.rs:43]: Hm, ... so I guess this is dead code now after these changes. However, I am not sure if removing this actually makes sense since `CacheCell` is a low-level API intended for general use, not only ink! internal use.

File: crates/storage/src/lib.rs
Comment (Robbepop) [inline crates/storage/src/lib.rs:23]: This PR is mergable as soon as this feature is stabilizied.

File: crates/storage/src/lazy/lazy_array.rs
Comment (Robbepop) [inline crates/storage/src/lazy/lazy_array.rs:228]: Whoat? Looks way more complicated than before and also seems to be allocating heap memory which is against the entire point of small vec and lazy array data structures. Please elaborate why.

**Review by Robbepop — CHANGES_REQUESTED** [review]

File: crates/storage/src/lazy/cache_cell.rs
Comment (cmichi) [reply crates/storage/src/lazy/cache_cell.rs]: Brought it back.

File: crates/storage/src/lazy/lazy_array.rs
Comment (cmichi) [reply crates/storage/src/lazy/lazy_array.rs]: Yeah… I pushed a different approach now, that one requires `#![feature(array_map)]` though 🙈.

Comment (Robbepop) [reply crates/storage/src/lazy/lazy_array.rs]: Nightly Rust features are a show stopper for us. Please look if the `array_init` crate can help you here. We are already depending on it anyway.

**Comment by cmichi** [comment]
@Robbepop Today I noticed that there already is a feature `const-generics` in the code for `array-init`. It does exactly what we need and I temporarily added it to this PR and migrated the code to use it. I talked to the maintainer and he plans to make it available by default as soon as `min_const_generics` are stabilized.

Yeah so overall we just need to wait until it's stabilized, remove the feature flags from this PR and are good to merge then.

I've marked the PR as blocked until then.

File: crates/storage/src/lib.rs
Comment (Robbepop) [inline crates/storage/src/lib.rs:24]: no longer needed ;)

File: crates/storage/Cargo.toml
Comment (cmichi) [inline crates/storage/Cargo.toml:28]: We still have to pass this feature, since this dependency will only make it available be default once `min_const_generics` is in a stable release.

Comment (Robbepop) [reply crates/storage/Cargo.toml:28]: hmmm good point.
best would be to sync `array-init` with us or vice versa.

File: crates/storage/src/lazy/lazy_array.rs
Comment (Robbepop) [inline crates/storage/src/lazy/lazy_array.rs]: oh, good we removed that one

**Comment by Robbepop** [comment]
We concluded to push this PR forward even though `min_const_generics` has not yet been stabilized.
Since we do not have a guarantee that `min_const_generics` is actually shipping with Rust version 1.51 in 25th of march we will put the `LazyArray` as well as everything that depends on it (namely `ink_storage::collections::SmallVec`) behind the new `ink-unstable` crate feature for the `ink_storage` crate.
This way users can decide themselves if they want to depend on this feature (as opt-in) and therefore risk being forced to nightly Rust for a longer time.

The big advantage is that as soon as `min_const_generics` is stabilized we can simply release the new `SmallVec` and `LazyArray` data structures without breaking user code.

File: crates/storage/src/lazy/cache_cell.rs
Comment (Robbepop) [inline crates/storage/src/lazy/cache_cell.rs:41]: what needed for?

**Review by Robbepop — APPROVED** [review]
LGTM :rocket:

**Pull request merged**
